File: decision_studio/reasoning/theory_value.py
"""Theories of value: coverage of options, and the decider's conviction.

## Conviction is not confidence

``Theory.confidence`` is the model's estimate, and everything in this codebase
treats it with suspicion: it is shown as a band, discounted by objections, and
never raised by a simulated experiment. *Conviction* is the decider's own degree
of belief that the theory holds — the quantity the theory-based view says a
decision actually turns on, and the one Bocconi's Aristotle asks for alongside
each causal map.

It is stated once, as a **prior**, and then moves only through things observed
in the world: a tripwire firing or not, a field experiment's result, a tested
link. Each observation is entered as a **likelihood ratio** — how much more
likely the observation is if the theory is true than if it is false — and the
update is Bayes' rule in odds form::

    posterior odds = prior odds x LR1 x LR2 x ...

Odds form makes the history replayable and order-independent, so the stored
rows are the record and the current conviction is always recomputed from them.

## Which evidence counts

Evidence recorded *after* the latest prior is applied to it. Evidence recorded
before it is shown and not applied: someone restating their conviction after
seeing a test has already taken the test into account, and applying it again
would count it twice. Evidence recorded before any prior waits, visibly, for
one.

## Likelihood ratios are elicited, with defaults

Asking for a likelihood ratio directly produces bad numbers, as asking for a
probability does. The UI offers five verbal steps (``LIKELIHOOD_SCALE``); a
tripwire supplies a default from its direction (``tripwire_likelihood``). Every
one is a hand-chosen, uncalibrated convention, like the thresholds in HANDOVER
§10 — exposed so it can be argued with.
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field
from typing import Any, Iterable
from uuid import UUID

from sqlalchemy import delete, select
from sqlalchemy.ext.asyncio import AsyncSession

from decision_studio.db.models import Theory, TheoryBelief

#: Bounds on any stated or computed conviction. Certainty cannot be updated —
#: odds of 0 or infinity absorb every likelihood ratio — so nobody may state it.
MIN_CONVICTION = 0.01
MAX_CONVICTION = 0.99

#: The verbal scale offered for a likelihood ratio. Symmetric in log-odds, so
#: "strongly for" and "strongly against" cancel exactly.
LIKELIHOOD_SCALE: dict[str, float] = {
    "strongly_for": 4.0,
    "for": 2.0,
    "neutral": 1.0,
    "against": 0.5,
    "strongly_against": 0.25,
}

#: Bounds on any likelihood ratio accepted from a client.
MIN_LR = 1 / 20
MAX_LR = 20.0

SOURCES = ("elicited", "tripwire", "field_experiment", "link_hypothesis")


def clamp_probability(value: float) -> float:
    return max(MIN_CONVICTION, min(MAX_CONVICTION, float(value)))


def clamp_lr(value: float) -> float:
    return max(MIN_LR, min(MAX_LR, float(value)))


def bayes_update(prior: float, likelihood_ratios: Iterable[float]) -> float:
    """Posterior probability from a prior and likelihood ratios, in odds form."""
    p = clamp_probability(prior)
    log_odds = math.log(p / (1 - p)) + sum(math.log(clamp_lr(lr)) for lr in likelihood_ratios)
    posterior = 1 / (1 + math.exp(-log_odds))
    return clamp_probability(posterior)


def tripwire_likelihood(direction: str, observed: bool) -> float:
    """The default likelihood ratio of a tripwire outcome.

    A falsifier that fires is strong evidence against (the user committed in
    advance that it would change their mind). One that does not fire is mild
    evidence for: the theory survived a test it could have failed, but absence
    of an observation is weaker than presence. A confirmer is the mirror image.
    """
    if direction == "confirms":
        return LIKELIHOOD_SCALE["strongly_for"] if observed else 1 / 1.5
    return LIKELIHOOD_SCALE["strongly_against"] if observed else 1.5


@dataclass
class ConvictionStep:
    """One piece of evidence and where it left the conviction."""

    id: str
    likelihood_ratio: float
    source: str
    source_id: str | None
    note: str | None
    created_at: Any
    applied: bool
    after: float | None


@dataclass
class Conviction:
    """A theory's conviction, replayed from its history."""

    theory_key: str
    prior: float | None = None
    prior_method: str | None = None
    prior_at: Any = None
    current: float | None = None
    steps: list[ConvictionStep] = field(default_factory=list)

    def as_dict(self) -> dict[str, Any]:
        return {
            "theory_key": self.theory_key,
            "prior": self.prior,
            "prior_method": self.prior_method,
            "prior_at": self.prior_at,
            "current": self.current,
            "steps": [step.__dict__ for step in self.steps],
        }


def replay(theory_key: str, rows: list[TheoryBelief]) -> Conviction:
    """Recompute a conviction from its stored rows, oldest first."""
    ordered = sorted(rows, key=lambda r: (r.created_at is None, r.created_at))
    priors = [r for r in ordered if r.kind == "prior" and r.value is not None]
    latest = priors[-1] if priors else None

    conviction = Conviction(theory_key=theory_key)
    if latest is not None:
        conviction.prior = latest.value
        conviction.prior_method = latest.method
        conviction.prior_at = latest.created_at
        conviction.current = latest.value

    for row in ordered:
        if row.kind != "evidence" or row.likelihood_ratio is None:
            continue
        applied = latest is not None and _after(row, latest)
        if applied:
            conviction.current = bayes_update(conviction.current, [row.likelihood_ratio])
        conviction.steps.append(ConvictionStep(
            id=str(row.id),
            likelihood_ratio=row.likelihood_ratio,
            source=row.source,
            source_id=str(row.source_id) if row.source_id else None,
            note=row.note,
            created_at=row.created_at,
            applied=applied,
            after=conviction.current if applied else None,
        ))
    return conviction


def _after(row: TheoryBelief, prior: TheoryBelief) -> bool:
    if row.created_at is None or prior.created_at is None:
        return True
    return row.created_at >= prior.created_at


async def convictions(
    session: AsyncSession, project_id: UUID, theory_keys: Iterable[UUID | str]
) -> dict[str, Conviction]:
    """Convictions for several theories, keyed by theory key."""
    keys = {UUID(str(k)) for k in theory_keys}
    if not keys:
        return {}
    rows = (await session.execute(
        select(TheoryBelief).where(
            TheoryBelief.project_id == project_id,
            TheoryBelief.theory_key.in_(keys),
        )
    )).scalars().all()
    grouped: dict[str, list[TheoryBelief]] = {str(k): [] for k in keys}
    for row in rows:
        grouped[str(row.theory_key)].append(row)
    return {key: replay(key, group) for key, group in grouped.items()}


async def _require_theory_key(session: AsyncSession, project_id: UUID, theory_key: UUID) -> None:
    found = await session.scalar(
        select(Theory.id).where(Theory.project_id == project_id, Theory.theory_key == theory_key).limit(1)
    )
    if found is None:
        raise LookupError(f"Theory {theory_key} not found in project")


async def state_prior(
    session: AsyncSession,
    project_id: UUID,
    theory_key: UUID,
    value: float,
    *,
    method: str = "direct",
    note: str | None = None,
) -> Conviction:
    """Record the decider's conviction before (further) evidence.

    Raises:
        LookupError: the theory is not in the project.
    """
    await _require_theory_key(session, project_id, theory_key)
    session.add(TheoryBelief(
        project_id=project_id,
        theory_key=theory_key,
        kind="prior",
        value=clamp_probability(value),
        source="elicited",
        method=method if method in ("lottery", "direct") else "direct",
        note=(note or "").strip()[:2000] or None,
    ))
    await session.commit()
    return (await convictions(session, project_id, [theory_key]))[str(theory_key)]


async def record_evidence(
    session: AsyncSession,
    project_id: UUID,
    theory_key: UUID,
    likelihood_ratio: float,
    *,
    source: str,
    source_id: UUID | None = None,
    note: str | None = None,
    commit: bool = True,
) -> None:
    """Record an observation's likelihood ratio against a theory.

    Never called for a synthetic experiment: simulated stakeholders agreeing is
    a measurement of the model, not of the world (see reasoning/experiments.py).
    """
    if source not in SOURCES or source == "elicited":
        raise ValueError(f"Evidence cannot come from '{source}'")
    if source_id is not None:
        # One observation, one piece of evidence. Correcting a recorded result
        # (a tripwire marked "happened" by mistake, a link re-tested) replaces
        # the earlier row; adding a second would count the same observation
        # twice and compound it into the conviction.
        await session.execute(
            delete(TheoryBelief).where(
                TheoryBelief.project_id == project_id,
                TheoryBelief.theory_key == theory_key,
                TheoryBelief.source == source,
                TheoryBelief.source_id == source_id,
            )
        )
    session.add(TheoryBelief(
        project_id=project_id,
        theory_key=theory_key,
        kind="evidence",
        likelihood_ratio=clamp_lr(likelihood_ratio),
        source=source,
        source_id=source_id,
        note=(note or "").strip()[:2000] or None,
    ))
    if commit:
        await session.commit()


def option_coverage(
    anchor: dict[str, Any] | None, theories: list[Any]
) -> list[dict[str, Any]]:
    """For each option, how many current theories argue for and against it.

    An option no theory addresses is the most important row: the graph is silent
    about one of the choices, and a recommendation between options where one was
    never examined is not a comparison.
    """
    rows = []
    for option in (anchor or {}).get("options", []):
        bound = [t for t in theories if getattr(t, "option_key", None) == option["key"]]
        rows.append({
            "key": option["key"],
            "label": option["label"],
            "achieves": sum(1 for t in bound if t.predicted_effect == "achieves"),
            "threatens": sum(1 for t in bound if t.predicted_effect == "threatens"),
            "unclear": sum(1 for t in bound if t.predicted_effect not in ("achieves", "threatens")),
            "reaching_outcome": sum(1 for t in bound if getattr(t, "reaches_outcome", False)),
        })
    return rows

File: frontend/src/lib/theoryValue.ts
/**
 * Theories of value on the client: conviction elicitation, the evidence scale,
 * option coverage and the API for conviction, link hypotheses and field results.
 *
 * Conviction is the decider's belief that a theory holds — not the model's
 * confidence. It is elicited with lottery comparisons rather than a slider:
 * "how sure are you, 0 to 1" is the question known to produce bad numbers,
 * while "would you rather bet on this, or on a draw you win 70% of the time"
 * is one people answer consistently. Four answers bisect the interval to about
 * ±3%, which is finer than anyone's belief actually is.
 */

import { apiGet, apiPost } from './api/client.ts'
import type { DecisionAnchor } from '../types/graph.ts'
import type { Theory } from '../types/reasoning.ts'

// --- Lottery elicitation ---

export const LOTTERY_STEPS = 4

export type LotteryChoice = 'theory' | 'draw' | 'same'

export interface LotteryState {
  lo: number
  hi: number
  answers: Array<{ p: number; choice: LotteryChoice }>
  done: boolean
}

export const startLottery = (): LotteryState => ({ lo: 0, hi: 1, answers: [], done: false })

/** The draw's winning probability to offer next. */
export const lotteryProbe = (state: LotteryState): number =>
  Math.round(((state.lo + state.hi) / 2) * 100) / 100

export function answerLottery(state: LotteryState, choice: LotteryChoice): LotteryState {
  if (state.done) return state
  const p = lotteryProbe(state)
  const answers = [...state.answers, { p, choice }]
  if (choice === 'same') return { lo: p, hi: p, answers, done: true }
  // Preferring the bet on the theory means believing it more likely than p.
  const next = choice === 'theory' ? { lo: p, hi: state.hi } : { lo: state.lo, hi: p }
  return { ...next, answers, done: answers.length >= LOTTERY_STEPS }
}

/** The elicited conviction, never 0 or 1: certainty cannot be updated. */
export function lotteryEstimate(state: LotteryState): number {
  const mid = (state.lo + state.hi) / 2
  return Math.min(0.99, Math.max(0.01, Math.round(mid * 100) / 100))
}

/** A readable record of the answers, stored with the prior so it can be audited. */
export function describeLottery(state: LotteryState): string {
  return state.answers
    .map((a) => `${Math.round(a.p * 100)}%: ${a.choice === 'theory' ? 'theory' : a.choice === 'draw' ? 'draw' : 'same'}`)
    .join(', ')
}

// --- Evidence scale (mirrors reasoning/theory_value.py) ---

export const LIKELIHOOD_SCALE = {
  strongly_for: 4,
  for: 2,
  neutral: 1,
  against: 0.5,
  strongly_against: 0.25,
} as const

export type LikelihoodKey = keyof typeof LIKELIHOOD_SCALE

/** The verbal step nearest a likelihood ratio, for showing history. */
export function likelihoodKey(ratio: number): LikelihoodKey {
  let best: LikelihoodKey = 'neutral'
  for (const key of Object.keys(LIKELIHOOD_SCALE) as LikelihoodKey[]) {
    if (Math.abs(Math.log(LIKELIHOOD_SCALE[key]) - Math.log(ratio)) <
        Math.abs(Math.log(LIKELIHOOD_SCALE[best]) - Math.log(ratio))) best = key
  }
  return best
}

// --- Option coverage (mirrors theory_value.option_coverage) ---

export interface OptionCoverageRow {
  key: string
  label: string
  achieves: number
  threatens: number
  unclear: number
  reachingOutcome: number
}

export function optionCoverage(anchor: DecisionAnchor | null | undefined, theories: Theory[]): OptionCoverageRow[] {
  return (anchor?.options ?? []).map((option) => {
    const bound = theories.filter((t) => t.optionKey === option.key)
    return {
      key: option.key,
      label: option.label,
      achieves: bound.filter((t) => t.predictedEffect === 'achieves').length,
      threatens: bound.filter((t) => t.predictedEffect === 'threatens').length,
      unclear: bound.filter((t) => t.predictedEffect !== 'achieves' && t.predictedEffect !== 'threatens').length,
      reachingOutcome: bound.filter((t) => t.reachesOutcome).length,
    }
  })
}

// --- API ---

export interface ConvictionStep {
  id: string
  likelihoodRatio: number
  source: 'tripwire' | 'field_experiment' | 'link_hypothesis' | string
  note: string | null
  createdAt: string | null
  /** False when recorded before the latest prior: already reflected in it. */
  applied: boolean
  after: number | null
}

export interface Conviction {
  theoryKey: string
  prior: number | null
  priorMethod: string | null
  current: number | null
  steps: ConvictionStep[]
}

interface ApiConviction {
  theory_key: string
  prior: number | null
  prior_method: string | null
  current: number | null
  steps: Array<{
    id: string; likelihood_ratio: number; source: string; note: string | null
    created_at: string | null; applied: boolean; after: number | null
  }>
}

const toConviction = (api: ApiConviction): Conviction => ({
  theoryKey: api.theory_key,
  prior: api.prior,
  priorMethod: api.prior_method,
  current: api.current,
  steps: api.steps.map((s) => ({
    id: s.id, likelihoodRatio: s.likelihood_ratio, source: s.source, note: s.note,
    createdAt: s.created_at, applied: s.applied, after: s.after,
  })),
})

const convictionPath = (projectId: string, theoryKey: string) =>
  `/api/v1/graph/${projectId}/theory-keys/${theoryKey}/conviction`

export async function fetchConviction(projectId: string, theoryKey: string): Promise<Conviction> {
  return toConviction(await apiGet<ApiConviction>(convictionPath(projectId, theoryKey)))
}

export async function statePrior(
  projectId: string, theoryKey: string, value: number, method: 'lottery' | 'direct', note?: string,
): Promise<Conviction> {
  return toConviction(
    await apiPost<ApiConviction>(convictionPath(projectId, theoryKey), { value, method, note }),
  )
}

export interface LinkHypothesis {
  id: string
  theoryKey: string
  edgeId: string
  statement: string
  refutedIf: string
  cheapestTest: string
  priority: number
  leverage: number
  uncertainty: number
  status: 'open' | 'held' | 'refuted' | 'inconclusive'
  observedNote: string | null
}

interface ApiHypothesis {
  id: string; theory_key: string; edge_id: string; statement: string; refuted_if: string
  cheapest_test: string; priority: number; leverage: number; uncertainty: number
  status: LinkHypothesis['status']; observed_note: string | null
}

const toHypothesis = (api: ApiHypothesis): LinkHypothesis => ({
  id: api.id, theoryKey: api.theory_key, edgeId: api.edge_id, statement: api.statement,
  refutedIf: api.refuted_if, cheapestTest: api.cheapest_test, priority: api.priority,
  leverage: api.leverage, uncertainty: api.uncertainty, status: api.status,
  observedNote: api.observed_note,
})

export async function fetchHypotheses(projectId: string, theoryKey: string): Promise<LinkHypothesis[]> {
  const res = await apiGet<{ hypotheses: ApiHypothesis[] }>(
    `/api/v1/graph/${projectId}/hypotheses?theory_key=${encodeURIComponent(theoryKey)}`,
  )
  return res.hypotheses.map(toHypothesis)
}

export async function proposeHypotheses(projectId: string, theoryId: string): Promise<LinkHypothesis[]> {
  const res = await apiPost<{ hypotheses: ApiHypothesis[] }>(
    `/api/v1/graph/${projectId}/theories/${theoryId}/hypotheses`, {},
  )
  return res.hypotheses.map(toHypothesis)
}

export async function recordHypothesisResult(
  projectId: string, hypothesisId: string, result: 'held' | 'refuted' | 'inconclusive',
  likelihoodRatio?: number, note?: string,
): Promise<LinkHypothesis> {
  return toHypothesis(await apiPost<ApiHypothesis>(
    `/api/v1/graph/${projectId}/hypotheses/${hypothesisId}/result`,
    { result, likelihood_ratio: likelihoodRatio, note },
  ))
}

// --- Field tests: the only experiment allowed to move conviction ---

export interface FieldTest {
  id: string
  hypothesis: string
  design: string
  measure: string | null
  costEstimate: string | null
  durationDays: number | null
  status: 'designed' | 'executed' | 'abandoned'
  summary: string | null
}

interface ApiExperiment {
  id: string; kind: 'synthetic' | 'field'; hypothesis: string; design: string
  measure: string | null; cost_estimate: string | null; duration_days: number | null
  status: FieldTest['status']; summary: string | null
}

const toFieldTest = (api: ApiExperiment): FieldTest => ({
  id: api.id, hypothesis: api.hypothesis, design: api.design, measure: api.measure,
  costEstimate: api.cost_estimate, durationDays: api.duration_days, status: api.status,
  summary: api.summary,
})

export async function fetchFieldTests(projectId: string, theoryId: string): Promise<FieldTest[]> {
  const res = await apiGet<{ experiments: ApiExperiment[] }>(
    `/api/v1/graph/${projectId}/experiments?theory_id=${encodeURIComponent(theoryId)}`,
  )
  return res.experiments.filter((e) => e.kind === 'field').map(toFieldTest)
}

export async function designFieldTest(projectId: string, theoryId: string): Promise<FieldTest> {
  return toFieldTest(await apiPost<ApiExperiment>(
    `/api/v1/graph/${projectId}/theories/${theoryId}/experiments/field`, {},
  ))
}

export async function recordFieldResult(
  projectId: string, experimentId: string, result: 'supports' | 'refutes' | 'inconclusive',
  note?: string,
): Promise<void> {
  await apiPost(`/api/v1/graph/${projectId}/experiments/${experimentId}/result`, { result, note })
}
